### PR TITLE
V3: Add more HTTP error types

### PR DESCRIPTION
**Pull request by pierre-emmanuelJ**
# Description
Add more HTTP error types to be more efficient on our api errors

## Checklist
(For exoscale contributors)

* [x] Changelog updated (under *Unreleased* block)
* [x] For a new resource or new attributes: test added/updated

Test:

```Go
func main() {
	c := credentials.NewEnvCredentials()
	cli, err := v3.NewClient(c)
	if err != nil {
		log.Fatal(err)
	}

	_, err = cli.GetInstance(context.Background(), "bad request not an uuid")
	if err != nil {
		if errors.Is(err, v3.ErrBadRequest) {
			log.Println("OK")
		}
	}

	_, err = cli.GetInstance(context.Background(), "e4a96b3f-9d50-4137-bb02-ba734c75a886")
	if err != nil {
		if errors.Is(err, v3.ErrNotFound) {
			log.Println("OK")
		}
	}
}

```

Output:

```
2024/07/11 16:47:39 OK
2024/07/11 16:47:39 OK
```

We will need to update CSI:
https://github.com/exoscale/exoscale-csi-driver/blob/main/driver/controller.go#L337
from `ErrInvalidRequest` to `ErrBadRequest` (the dedicated code to the returned error)

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -8,14 +8,12 @@ require (
...
-require (
-	github.com/kr/pretty v0.3.1 // indirect
-	gopkg.in/yaml.v3 v3.0.1 // indirect
-)
+require gopkg.in/yaml.v3 v3.0.1 // indirect
 
 require (
 	github.com/davecgh/go-spew v1.1.2-0.20180830191138-d8f796af33cc // indirect
 	github.com/hashicorp/go-cleanhttp v0.5.2 // indirect
+	github.com/kr/pretty v0.3.1 // indirect
```
Comment (sauterp) [inline go.mod:16]: the changes in this file look like noise to me, how did they come about?

File: v3/errors.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+			if err = json.Unmarshal(data, &res); err != nil {
+				return fmt.Errorf("error unmarshaling response: %s", err)
+			}
+		} else {
+			res.Message = string(data)
+		}
+
+		err, ok := httpStatusCodeErrors[resp.StatusCode]
+		if ok {
+			if err == ErrNotFound {
```
Comment (sauterp) [inline v3/errors.go:117]: Why do we need this special case?

**Review by sauterp — APPROVED** [review]

File: v3/errors.go (hunk 2 above)
Comment (pierre-emmanuelJ) [reply v3/errors.go:117]: I reproduced the past behavior, to not ship the message (probably no message for 404 that why) in the error for 404 but I think I can test to do like other codes, yes

Comment (pierre-emmanuelJ) [reply v3/errors.go]: I removed it message is here and more explicit then, thanks @sauterp

File: v3/errors.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+)
+
+var (
+	ErrBadRequest                    = fmt.Errorf(http.StatusText(http.StatusBadRequest))
```
Comment (aureliar8) [inline v3/errors.go:11]:
```suggestion
	ErrBadRequest                    = errors.New((http.StatusText(http.StatusBadRequest))
```

Comment (pierre-emmanuelJ) [reply v3/errors.go]: https://github.com/exoscale/egoscale/pull/641/commits/4a5dab3dafa22d32955f76dae394378f17177e2e

**Review by aureliar8 — APPROVED** [review]

**Review by aureliar8 — COMMENTED** [review]
Just note that it's a breaking change. 

I'm not sure what's the current policy for egoscale v3

**Pull request merged**